woocommerce/woocommerce
# [Accessibility] Focus coupon input if it has an error

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




Closes #37602 .

In order to create small PRs, the issue above will be fixed along two PRs.
1. This PR is dedicated to making the coupon input accessible when there is an error after sending it on:
    - Classic Cart page.
    - Block-based Cart page.
    - Classic Checkout page.
    - Block-based Checkout page.
2.  [[Accessibility] Improve checkout error messages](https://github.com/woocommerce/woocommerce/pull/49094)

The issue associated with this PR reports more accessibility problems, but most of them have already been fixed by other PRs. Such as:
- [x] Add `aria-invalid="true"` to invalid form fields and update to `aria-invalid="false"` or remove for valid fields.
- [x] Add error messages nested inside the form field.
- [x] On the checkout page, focus the first field with an error.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:


**Note**: I noticed that when the Storefront theme is installed, these CSS files aren't loaded on the classic Cart and Checkout pages:
  - `plugins/woocommerce/client/legacy/css/woocommerce-smallscreen.scss`
  - `plugins/woocommerce/client/legacy/css/woocommerce.scss`
This means that the coupon inline error style will be different depending on the theme installed.

1. On both the Cart and Checkout pages, add a coupon code that doesn't exist.
3. Verify the error message is added below the input. "Coupon "..." does not exist!"
4. Verify the focus is moved to the coupon input.
5. Verify the error message disappears after moving the focus away from the input or as soon as you type any new value to the input.
6. Repeat this process on both the classic and the block-based pages.
7. Test the classic cart against the following themes (check [this comment](https://github.com/woocommerce/woocommerce/pull/48738#pullrequestreview-2204533147) for more context):
    - Twenty Seventeen
    - Twenty Nineteen
    - Twenty Twenty
    - Twenty Twenty-One
    - Twenty Twenty-Two
    - Twenty Twenty-Three
    - Twenty Twenty-Four



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
